ethcore/src/engines/clique/mod.rs

```diff
@@ -256,93 +255,104 @@ impl Clique {
+				break;
+			}
 
-				let last_checkpoint_state = match block_state_by_hash.get_mut(&last_checkpoint_hash) {
-					Some(state) => state.clone(),
-					None => self.new_checkpoint_state(&last_checkpoint_header)?,
-				};
+			let query = BlockId::Hash(last_parent_hash);
+			let next_header = client.block_header(query).ok_or_else(|| BlockError::UnknownParent(last_parent_hash))?;
+			chain.push_front(next_header.decode()?);
+		}
 
-				block_state_by_hash.insert(last_checkpoint_header.hash(), last_checkpoint_state.clone());
+		// Catching up state, note that we don't really store block state for intermediary blocks for speed.
+		let backfill_start = time::Instant::now();
+		trace!(target: "engine",
+			   "Back-filling block state. last_checkpoint_number: {}, target: {}({}).",
+			   last_checkpoint_number, block_number, block_hash);
 
-				// Backfill!
-				let mut new_state = last_checkpoint_state.clone();
-				for item in chain {
-					new_state.apply(item, false)?;
-				}
-				new_state.calc_next_timestamp(header.timestamp(), self.period)?;
-				block_state_by_hash.insert(header.hash(), new_state.clone());
+		// Get the state for last checkpoint.
+		let last_checkpoint_hash = *chain.front()
+			.expect("chain has at least one element; qed")
+			.parent_hash();
 
-				let elapsed = backfill_start.elapsed();
-				trace!(target: "engine",
-						"Back-filling succeed, took {} ms.",
-						// replace with Duration::as_millis after rust 1.33
-						elapsed.as_secs() as u128 * 1000 + elapsed.subsec_millis() as u128,
-				);
+		let query = BlockId::Hash(last_checkpoint_hash);
+		let last_checkpoint_header = client.block_header(query)
+			.ok_or_else(|| EngineError::CliqueMissingCheckpoint(last_checkpoint_hash))?.decode()?;
 
-				Ok(new_state)
-			}
+		let last_checkpoint_state = match block_state_by_hash.get_mut(&last_checkpoint_hash) {
+			Some(state) => state.clone(),
+			None => self.new_checkpoint_state(&last_checkpoint_header)?,
+		};
+
+		block_state_by_hash.insert(last_checkpoint_header.hash(), last_checkpoint_state.clone());
+
+		// Backfill!
+		let mut new_state = last_checkpoint_state.clone();
+		for item in chain {
+			new_state.apply(item, false)?;
 		}
+		new_state.calc_next_timestamp(header.timestamp(), self.period)?;
+		block_state_by_hash.insert(block_hash, new_state.clone());
+
+		let elapsed = backfill_start.elapsed();
+		trace!(target: "engine",
+			   "Back-filling succeed, took {} ms.",
+			   // replace with Duration::as_millis after rust 1.33
+			   elapsed.as_secs() as u128 * 1000 + elapsed.subsec_millis() as u128,
+		);
+
+		Ok(new_state)
+	}
+
+	pub fn find_header(&self, query: BlockId) -> Result<Header, Error> {
+		let client = self.client.read().as_ref().and_then(|w| w.upgrade()).ok_or(EngineError::RequiresClient)?;
+		let header = client.block_header(query).ok_or_else(|| "block not found")?.decode()?;
+		Ok(header)
+	}
+
+	pub fn get_signers(&self, query: BlockId) -> Result<Vec<Address>, Error> {
```

Comment:
I think this method should wrap `state` (https://github.com/goerli/parity-goerli/blob/master/ethcore/src/engines/clique/mod.rs#L258) to get the signers for a given block hash.  The full signers list is only present in the header at the epoch transition.

Reply:
done